pkg/kubelet/userns_manager.go

```diff
@@ -0,0 +1,476 @@
+	for i := b.firstIndex; i < len(b.data); i++ {
+		if b.data[i] == 0xFFFFFFFF {
+			continue
+		}
+		for j := uint32(0); j < bitsUint32; j++ {
+			if (b.data[i]>>j)&0x1 == 0 {
+				v := uint32(i)*bitsUint32 + j
+				b.set(v)
+				b.firstIndex = int(i)
+				return v, nil
+			}
+		}
+	}
+	return 0, fmt.Errorf("could not find an empty slot")
+}
+
+func (b *bitArray) clear(index uint32) {
+	i := index / bitsUint32
+	if i < uint32(b.firstIndex) {
+		b.firstIndex = int(i)
+	}
+	b.data[i] &= ^(1 << (index % bitsUint32))
+}
+
+// length for the user namespace to create (65536).
+const userNsLength = math.MaxUint16 + 1
+
+// Limit the total number of pods using userns in this node to this value.
+// This is an alpha limitation that will probably be lifted later.
+const maxPods = 1024
+
+type usernsManager struct {
+	used     *bitArray
+	usedBy   map[string]uint32
+	removed  int
+	currPods int
+	kl       *Kubelet
+	sync.Mutex
+}
+
+// UserNamespace holds the configuration for the user namespace.
+type userNamespace struct {
+	// UIDs mappings for the user namespace.
+	UIDMappings []idMapping `json:"uidMappings"`
+	// GIDs mappings for the user namespace.
+	GIDMappings []idMapping `json:"gidMappings"`
+}
+
+// Pod user namespace mapping
+type idMapping struct {
+	// Required.
+	HostId uint32 `json:"hostId"`
+	// Required.
+	ContainerId uint32 `json:"containerId"`
+	// Required.
+	Length uint32 `json:"length"`
+}
+
+func MakeUserNsManager(kl *Kubelet) (*usernsManager, error) {
+	m := usernsManager{
+		// Create a bitArray for all the UID space (2^32).
+		// As a by product of that, no index param to bitArray can be out of bounds (index is uint32).
+		used:   makeBitArray((math.MaxUint32 + 1) / userNsLength),
+		usedBy: make(map[string]uint32),
+		kl:     kl,
+	}
+	// First block is reserved for the host.
+	m.used.set(0)
+
+	// Second block will be used for phase II. Don't assign that range for now.
+	m.used.set(1)
+
+	if kl != nil {
+		found, err := kl.listPodsFromDisk()
+		if err != nil {
+			return nil, fmt.Errorf("user namespace manager can't read pods from disk: %w", err)
+		}
+
+		for _, uid := range found {
+			if err := m.recordPodMappings(uid); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return &m, nil
+}
+
+// recordPodMappings registers the range used for the user namespace iif the
+// usernsConfFile exits in the pod directory.
+func (m *usernsManager) recordPodMappings(pod types.UID) error {
+	usernsConfFile := m.getUserNamespaceMappingsFile(pod)
+
+	// If the size is too big, something is fishy. Don't read it all in mem.
+	// Ignore errors, the file might not be created yet and afterwards we
+	// check which error was returned.
+	if info, err := os.Stat(usernsConfFile); err == nil && info.Size() >= 1024*1024 {
+		return fmt.Errorf("pod user namespace file is too big: %q", usernsConfFile)
+	}
+
+	content, err := os.ReadFile(usernsConfFile)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if string(content) == "" {
```

Comment:
I'd comment this to explain what's being tested for (and why the early exit here never represents an error).

Reply:
Yeah, can we add a comment here for the case where we expect this to be empty and not return an error?